src/Compute/Metadata/GCEMetadataStream.php

```diff
@@ -0,0 +1,173 @@
+
+    /**
+     * The current seek position.
+     */
+    private $position;
+
+    /**
+     * The metadata value.
+     */
+    private $val;
+
+    /**
+     * The metadata reader.
+     */
+    private $reader;
+
+    /**
+     * A static method to register this stream wrapper.
+     */
+    public static function register()
+    {
+        stream_wrapper_register(
+            'gce-metadata',
+            '\Google\Cloud\Compute\Metadata\GCEMetadataStream'
+        )
+            or die('Failed to register GCEMetadataStream');
+    }
+
+    public function __construct()
+    {
+        $this->reader = new StreamReader();
+    }
+
+    /**
+     * A method to replace the reader implementation.
+     */
+    public function setReader($reader)
+    {
+        $this->reader = $reader;
+    }
+
+    // @codingStandardsIgnoreStart
+
+    // These methods are mandatory for implementing stream wrappers,
+    // so marking them with ignore tag.
+    //
+    // TODO: Find a way to just suppress certain types of errors with
+    // phpcs.
+    public function stream_open($path, $mode, $options, &$opened_path)
+    {
+        $url = parse_url($path);
+        if ($url['host'] != 'project' and $url['host'] != 'instance') {
+            return false;
+        }
+        $metadata_url = self::METADATA_BASE_URL.$url['host'].'/attributes'
+            .$url['path'];
+        $this->val = $this->reader->read($metadata_url);
+        $this->position = 0;
+
+        return true;
+    }
+
+    public function stream_stat()
+    {
+        return array();
+    }
+
+    public function stream_read($count)
+    {
+        $ret = substr($this->val, $this->position, $count);
+        $this->position += strlen($ret);
+
+        return $ret;
+    }
+
+    public function stream_write($data)
+    {
+        throw new \Exception('stream_write is not implemented.');
+    }
+
+    public function stream_tell()
+    {
+        return $this->position;
+    }
+
+    public function stream_eof()
+    {
+        return $this->position >= strlen($this->val);
+    }
+
+    public function stream_seek($offset, $whence)
+    {
+        switch ($whence) {
+            case SEEK_SET:
+                if ($offset < strlen($this->val) && $offset >= 0) {
+                    $this->position = $offset;
+
+                    return true;
+                } else {
+                    return false;
+                }
+                break;
+
+            case SEEK_CUR:
+                if ($offset >= 0) {
+                    $this->position += $offset;
+
+                    return true;
+                } else {
+                    return false;
+                }
+                break;
+
+            case SEEK_END:
+                if (strlen($this->val) + $offset >= 0) {
+                    $this->position = strlen($this->val) + $offset;
+
+                    return true;
+                } else {
+                    return false;
+                }
+                break;
```

Comment:
I'd rather see this for each return statement:

```php
if (strlen($this->val) + $offset >= 0) {
   $this->position = strlen($this->val) + $offset;

   return true;
}

return false;
```

It's clearer to read and the logic is more straightforward. The `breaks` will never be executed and the `else` is unnecessary.